pkg/filesystem/fs.go

```diff
@@ -352,61 +350,66 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string, s *stor
 	case config.FsDriverFscache:
 		err = fs.mountRemote(fsManager, useSharedDaemon, d, rafs)
 		if err != nil {
-			return errors.Wrapf(err, "mount file system by daemon %s, snapshot %s", d.ID(), snapshotID)
+			err = errors.Wrapf(err, "mount file system by daemon %s, snapshot %s", d.ID(), snapshotID)
 		}
 	case config.FsDriverFusedev:
 		err = fs.mountRemote(fsManager, useSharedDaemon, d, rafs)
 		if err != nil {
-			return errors.Wrapf(err, "mount file system by daemon %s, snapshot %s", d.ID(), snapshotID)
+			err = errors.Wrapf(err, "mount file system by daemon %s, snapshot %s", d.ID(), snapshotID)
 		}
 	case config.FsDriverBlockdev:
 		err = fs.tarfsMgr.MountTarErofs(snapshotID, s, rafs)
 		if err != nil {
-			return errors.Wrapf(err, "mount tarfs for snapshot %s", snapshotID)
+			err = errors.Wrapf(err, "mount tarfs for snapshot %s", snapshotID)
 		}
 	default:
-		return errors.Errorf("unknown filesystem driver %s for snapshot %s", fsDriver, snapshotID)
+		err = errors.Errorf("unknown filesystem driver %s for snapshot %s", fsDriver, snapshotID)
 	}
 
 	// Persist it after associate instance after all the states are calculated.
-	if err := fsManager.AddRafsInstance(rafs); err != nil {
-		return errors.Wrapf(err, "create instance %s", snapshotID)
+	if err == nil {
+		if err := fsManager.AddRafsInstance(rafs); err != nil {
+			return errors.Wrapf(err, "create instance %s", snapshotID)
+		}
+	}
+
+	if err != nil {
+		_ = fs.Umount(ctx, snapshotID)
```

Comment:
I think we'd better ensure the `umounted` state is always true when `mount` error happens, instead of handling it outside `mount` func.

Reply:
It's a hard situation, error happens on error handling code path. Any suggestion about how to handling such a case?

Reply:
I'm reusing `Umount()` to do error recover, do you mean I should not call `Umount()` here?

Reply:
Yes, how about putting `defer Umount()` into `mountRemote()` and `MountTarErofs()`? The returned error of these funcs indicates that the `mount` operation has failed, it should be `umounted` state now. But I think maybe it's not a blocker for this PR.